metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationZkWriter.java

```diff
@@ -128,23 +133,55 @@ public void visitPartition(TopicIdPartition topicIdPartition, PartitionRegistrat
                         return; // topic deleted in KRaft
                     }
 
+                    // If there is failure in previous Zk writes, We could end up with Zookeeper
+                    // containing with partial or without any partitions for existing topics. So
+                    // accumulate the partition ids to check for any missing partitions in Zk.
+                    partitionsInZk
+                        .computeIfAbsent(topic.id(), __ -> new HashSet<>())
+                        .add(topicIdPartition.partition());
+
                     // Check if the KRaft partition state changed
                     PartitionRegistration kraftPartition = topic.partitions().get(topicIdPartition.partition());
                     if (!kraftPartition.equals(partitionRegistration)) {
                         changedPartitions.computeIfAbsent(topicIdPartition.topicId(), __ -> new HashMap<>())
                             .put(topicIdPartition.partition(), kraftPartition);
                     }
+
+                    // Check if partition assignment has changed. This will need topic update.
+                    if (!kraftPartition.hasSameAssignment(partitionRegistration)) {
+                        changedTopics.add(topic.id());
+                    }
                 }
             });
 
-        createdTopics.forEach(topicId -> {
+        // Check for any partition changes in existing topics.
+        for (Uuid topicId : topicsInZk) {
+            TopicImage topic = topicsImage.getTopic(topicId);
+            Set<Integer> topicPartitionsInZk = partitionsInZk.getOrDefault(topicId, new HashSet<>());
+            if (!topicPartitionsInZk.equals(topic.partitions().keySet())) {
+                Map<Integer, PartitionRegistration> newTopicPartitions = new HashMap<>(topic.partitions());
+                topicPartitionsInZk.forEach(newTopicPartitions::remove);
+                newPartitions.put(topicId, newTopicPartitions);
+                changedTopics.add(topicId);
```

Comment:
I think it's possible that we can see a difference between `topicPartitionsInZk` (computed based on the partition znodes) and the topic assignment in Kraft, but actually have a correct assignment in ZK (stored in the topic znode). That said, I don't think there's any harm in re-writing the kraft assignment to ZK.